In [82]:
import cv2
from sklearn.cluster import KMeans
from keras.applications.vgg16 import VGG16, preprocess_input
from tensorflow.keras.preprocessing.image import img_to_array
import numpy as np
import pandas as pd
import os
from math import floor,sqrt
from moviepy.editor import VideoFileClip
import librosa
from sklearn.preprocessing import StandardScaler, LabelEncoder
import h5py
from sklearn.metrics import precision_score, recall_score, f1_score
import objectDetection as od


## Video Summarization

### Annotations loader

In [83]:
# Load annotations
def load_annotations(anno_file, info_file):
    # Read the annotation and info files
    annotations = pd.read_csv(anno_file, sep='\t', header=None)
    info = pd.read_csv(info_file, sep='\t', header=None)
    
    # Rename columns for better understanding
    annotations.columns = ['video_id', 'category', 'importance_score']
    info.columns = ['category_code', 'video_id', 'title', 'url', 'length']
    
    # print( annotations, info)
    
    return annotations, info

### Preprocessing and Feature Extraction


#### a. Video Processing and Frame Extraction


In [84]:
def extract_frames(video_path, frame_rate=1):
    video = cv2.VideoCapture(video_path)
    count = 0
    success = True
    frames = []
    
    while success:
        success, image = video.read()
        if count % frame_rate == 0 and success:
            frames.append(image)
        count += 1

    video.release()
    return frames

#### b. Audio Extraction


In [85]:

def extract_audio_from_video(video_path, output_audio_path):
    video = VideoFileClip(video_path)
    audio = video.audio
    audio.write_audiofile(output_audio_path)
    video.close()

#### c.Audio Features

In [86]:
def extract_audio_features_for_each_frame(audio_path, frame_rate):
    y, sr = librosa.load(audio_path)

    # Calculate the number of audio samples per video frame
    samples_per_frame = sr / frame_rate

    # Initialize an array to store MFCCs for each frame
    mfccs_per_frame = []

    # Iterate over each frame and extract corresponding MFCCs
    for frame in range(int(len(y) / samples_per_frame)):
        start_sample = int(frame * samples_per_frame)
        end_sample = int((frame + 1) * samples_per_frame)

        # Ensure the end sample does not exceed the audio length
        end_sample = min(end_sample, len(y))

        # Extract MFCCs for the current frame's audio segment
        mfccs_current_frame = librosa.feature.mfcc(y=y[start_sample:end_sample], sr=sr, n_mfcc=13)
        mfccs_processed = np.mean(mfccs_current_frame.T, axis=0)
        mfccs_per_frame.append(mfccs_processed)
    print(np.array(mfccs_per_frame).shape)

    return mfccs_per_frame[:len(mfccs_per_frame)-1]


#### d. Feature Extraction (example with visual features using a CNN)


In [87]:
model = VGG16(weights='imagenet', include_top=False)

def extract_visual_features(frames):
    features = []
    for frame in frames:
        if frame is not None:
            img = cv2.resize(frame, (224, 224))  # Resize frame to 224x224
            img = img_to_array(img)        # Convert to array
            img = np.expand_dims(img, axis=0)    # Add batch dimension
            img = preprocess_input(img)          # Preprocess for VGG16
            
            feature = model.predict(img,use_multiprocessing=True,workers=4)
            features.append(feature.flatten())

    return features

### Connect Audio/Annotation/Video

#### Annotation To List

In [88]:
def annotation2List(annotation_features):
    # Make the string '1,1,1,3,2,2,4,4,1' to float list
    annotation_float_array=[]
    for annotation in annotation_features:
        
        if isinstance(annotation,str):
            annotation = annotation.split(',')
        if isinstance(annotation,list):
            for anno in annotation:
                annotation_float_array.append(float(anno))
        else:
            annotation_float_array.append(float(annotation))
    return annotation_float_array

#### Extract Data

In [89]:
import pickle
with open('objects.pkl', 'rb') as f:
    objects = pickle.load(f)

with open('frames.pkl', 'rb') as f:
    frames = pickle.load(f)

with open('labels.pkl', 'rb') as f:
    labels = pickle.load(f)

In [90]:
def extractData(video_path, anno_file, info_file):
    # Extract frames from the video
    frames = extract_frames(video_path)
    # print('frames',len(frames)) 

    # # Extract visual features
    visual_features = extract_visual_features(frames) #type: ignore

    # Extract audio
    audio_output_path = 'datasets/extractedAudio/extracted_audio.wav'
    extract_audio_from_video(video_path, audio_output_path) 

    # Extract audio features
    audio_features = extract_audio_features_for_each_frame(audio_output_path,30)

    # # Load annotations
    # annotations, info = load_annotations(anno_file, info_file)
    
    return visual_features, frames,audio_features#type: ignore


#### Kmeans and feature connection

##### One Hot Encoding

In [91]:
def one_hot_encode(objects_in_frame, unique_objects):
    """
    Convert a list of objects detected in a frame to a one-hot encoded vector.
    
    Args:
    objects_in_frame: List of objects detected in a frame.
    unique_objects: List of all unique objects across all frames.

    Returns:
    One-hot encoded vector representing the presence of objects in the frame.
    """
    encoding = [1 if obj in objects_in_frame else 0 for obj in unique_objects]
    return np.array(encoding)


##### Kmeans

In [92]:
def MultiModalKMeans(video_path, anno_file, info_file, num_clusters=None):
    """
    Integrate visual, audio, and annotation features from a video,
    and perform clustering on the combined features.

    :param video_path: Path to the video file.
    :param anno_file: Path to the annotation file.
    :param info_file: Path to the info file.
    :param num_clusters: Number of clusters to use in KMeans.
    :return: Cluster labels for each data point.
    """
    
    # Extract data from video
    visual_features,frames,audio_features=extractData(video_path, anno_file, info_file)
        
    objects = od.detect_objects_in_all_frames(frames,yolo_model,classes) #type: ignore
            
    unique_objects = sorted(list(set([obj for frame_objects in objects for obj in frame_objects]))) #type: ignore

    encoded_objects = [one_hot_encode(frame_objects, unique_objects) for frame_objects in objects] #type: ignore

    # Combine features with padding
    combined_features = []

    for i, frame in enumerate(frames):
        
        # annotation_float_array = annotation2List(annotation_features[i])
        
        combined_feature = np.concatenate([
            np.array(visual_features[i], dtype=float),
            np.array(audio_features[i], dtype=float),
            np.array(encoded_objects[i], dtype=float),
        ])
        combined_features.append(combined_feature)

    # Convert to 2D NumPy array and normalize
    combined_features_array = np.array(combined_features)
    
    combined_features_normalized = StandardScaler().fit_transform(combined_features_array)
         
    print("Number of clusters:", num_clusters)
    
    # Perform clustering
    print("Performing clustering...")
    kmeans = KMeans(n_clusters=num_clusters) #type: ignore
    kmeans.fit(combined_features_normalized)
    
    return [frames, kmeans.labels_]
    # return the frames and the labels


### Summarization

#### Find min cluster

In [93]:
from collections import Counter
def getMinClusster(labels):
    cluster_counts = Counter(labels)

    # Find the cluster with the maximum number of frames
    return min(cluster_counts, key=cluster_counts.get),cluster_counts #type: ignore

#### Frames Selection

In [156]:
def frame_selection(frames, labels):
    """
    Create a summary by selecting frames from the most populous cluster.

    :param frames: List of frames.
    :param labels: Cluster labels for each frame.
    :return: List of frames belonging to the most populous cluster.
    """
    # Count the number of frames in each cluster
    max_cluster,cluster_counts = getMinClusster(labels)

    # Initialize lists for indices and frames
    summary_indices = []  # to measure the importance based on annotation
    summary_frames = []

    # Iterate and select frames and their indices belonging to the most populous cluster
    for index, (frame, label) in enumerate(zip(frames, labels)):
        if label == max_cluster:
            summary_indices.append(index)
            summary_frames.append(frame)

    print(f"Number of frames selected for summary: {len(summary_frames)}, Cluster: {max_cluster}")
    # print all clusters len
    print(f"Cluster counts: {cluster_counts}")
    return summary_frames,summary_indices

def frameSelectionForEachCluster(frames):
    """
    Create a summary by selecting frames from the most populous cluster.

    :param frames: List of frames.
    :param labels: Cluster labels for each frame.
    :return: List of frames belonging to the most populous cluster.
    """
    # Initialize lists for indices and frames
    summary_indices = []  # to measure the importance based on annotation
    summary_frames = []

    # Iterate and select frames and their indices belonging to the most populous cluster
    for index, (frame) in enumerate(frames):
        summary_indices.append(index)
        summary_frames.append(frame)

    print(f"Number of frames selected for summary: {len(summary_frames)}")
    # print all clusters len
    # print(f"Cluster counts: {cluster_counts}")
    return summary_frames,summary_indices

#### Video Creator

In [95]:
def create_video_from_frames(frames, output_path, frame_rate=30):
    if not frames:
        print("No frames to create a video.")
        return 
    # Determine the width and height from the first frame
    height, width, layers = frames[0].shape

    # Define the codec and create VideoWriter object
    fourcc = cv2.VideoWriter_fourcc(*'mp4v')#type: ignore
    out = cv2.VideoWriter(output_path, fourcc, frame_rate, (width, height))#type: ignore

    # Write each frame to the video
    for frame in frames:
        out.write(frame)

    # Release the VideoWriter object
    out.release()

#### Load .Mat file

In [96]:
def decode_titles(encoded_titles, hdf5_file):
    decoded_titles = []
    for ref_array in encoded_titles:
        # Handle the case where each ref_array might contain multiple references
        for ref in ref_array:
            # Dereference each HDF5 object reference to get the actual data
            title_data = hdf5_file[ref]
            # Decode the title
            decoded_title = ''.join(chr(char[0]) for char in title_data)
            decoded_titles.append(decoded_title)
    return decoded_titles


def load_mat_file(file_path,videoID):
    """
    Load a .mat file and return its contents.

    :param file_path: Path to the .mat file.
    :return: Contents of the .mat file.
    """
    with h5py.File(file_path, 'r') as file:
        user_anno_refs=file['tvsum50']['user_anno'][:] # type: ignore
        video_refs=file['tvsum50']['video'][:] # type: ignore

        decoded_videos = decode_titles(video_refs,file)
    
        annotations = []        
        # Get the index from decoded video list to find the annotation for the video
        index = [i for i, x in enumerate(decoded_videos) if x.lower() in videoID.lower()][0]
        
        # Iterate over each reference
        for ref in user_anno_refs: # type: ignore
            # Dereference each HDF5 object reference
            ref_data = file[ref[0]]

            # Convert to NumPy array and add to the annotations list
            annotations.append(np.array(ref_data))
            
        return annotations[index]

#### f1score with ground_truth

In [168]:
def evaluate_frame_selection(ground_truth, summary_indices):
    """
    Evaluate the selected frames by comparing them with the ground truth.
    
    Args:
    ground_truth: Ground truth annotations.
    summary_indices: Indices of the selected frames.
    
    Returns:
    Average importance score, max importance score, and proportion of frames with high importance score.
    """
    
    # Evaluate the selected frames
    selected_importance_scores = ground_truth[summary_indices]
    
    # Calculate metrics
    if selected_importance_scores.size == 0:
        average_importance = 0
        max_importance = 0
        proportion_high_importance = 0
    else:
        average_importance = np.mean(selected_importance_scores)  # Average importance score
        max_importance = np.max(selected_importance_scores)  # Max importance score
        # Calculate the proportion of frames with high importance score
        proportion_high_importance = np.mean(selected_importance_scores >= np.floor(max_importance))
        
    return average_importance, max_importance,proportion_high_importance


In [204]:
def Evaluation(ground_truth_path,summary_indices,videoID):
    
    # Get the ground_truth
    ground_truth = np.array(load_mat_file(ground_truth_path, videoID))
    
    # Find the mean of all annotators
    gold_standard = np.mean(ground_truth, axis=0)
    
    # based on annotators find the avg_importance, max_importance, prop_high_importance
    avg_importance, max_importance, prop_high_importance = evaluate_frame_selection(gold_standard, summary_indices)

    # Thresholding based on max_importance
    threshold=floor(np.mean(max_importance))
    
    
    # Binary conversion
    binary_ground_truth = np.where(gold_standard >= threshold, 1, 0) 
    
    # Selected frames binary conversion
    selected_frames_binary = np.zeros_like(binary_ground_truth)
    selected_frames_binary[summary_indices] = 1
    
    # Calculate metrics
    precision = precision_score(binary_ground_truth, selected_frames_binary)
    recall = recall_score(binary_ground_truth, selected_frames_binary)
    f1 = f1_score(binary_ground_truth, selected_frames_binary, average='binary')
    
    # return metrics
    return threshold,precision,recall,f1,avg_importance,max_importance,prop_high_importance


#### KnapSack

In [99]:
def knapsack_for_video_summary(values, weights, capacity, scale_factor=30):
    """
    Apply the 0/1 Knapsack algorithm to select video segments for summarization.

    :param values: List of importance scores for each segment.
    :param weights: List of durations for each segment in seconds.
    :param capacity: Maximum total duration for the summary in seconds.
    :param scale_factor: Factor to scale weights to integers.
    :return: Indices of the segments to include in the summary.
    """
    # Scale weights and capacity
    weights = [int(w * scale_factor) for w in weights]
    capacity = int(capacity * scale_factor)

    n = len(values)
    K = [[0 for _ in range(capacity + 1)] for _ in range(n + 1)]

    # Build table K[][] in a bottom-up manner
    for i in range(n + 1):
        for w in range(capacity + 1):
            if i == 0 or w == 0:
                K[i][w] = 0
            elif weights[i-1] <= w:
                K[i][w] = max(values[i-1] + K[i-1][w-weights[i-1]], K[i-1][w])
            else:
                K[i][w] = K[i-1][w]

    # Find the selected segments
    res = K[n][capacity]
    w = capacity
    selected_indices = []

    for i in range(n, 0, -1):
        if res <= 0:
            break
        if res == K[i-1][w]:
            continue
        else:
            selected_indices.append(i-1)
            res = res - values[i-1]
            w = w - weights[i-1]

    selected_indices.reverse()
    return selected_indices


### importance score

In [166]:
def calculate_importance_scores(labels, detected_objects):
    """Here the importance for all the labels together..

    Args:
        labels (_type_): _description_
        detected_objects (_type_): _description_

    Returns:
        _type_: _description_
    """
    # Count the frequency of each label
    label_counts = Counter(labels)

    # Assign a basic importance score based on the inverse frequency of cluster labels
    base_importance_scores = [1 / label_counts[label] for label in labels]

    # Additional importance based on detected objects
    object_importance_scores = []
    for objects in detected_objects:
        if objects:  # if the list is not empty
            # Add extra importance for each detected object
            # You can customize this part based on the type and number of objects
            object_importance_scores.append(len(objects))
        else:
            object_importance_scores.append(0)

    # Combine base importance scores with object importance scores
    # Normalize object importance scores for simplicity
    max_object_score = max(object_importance_scores) if object_importance_scores else 1
    normalized_object_scores = [score / max_object_score for score in object_importance_scores]

    # Final importance score is a combination of base and object scores
    importance_scores = [base + obj for base, obj in zip(base_importance_scores, normalized_object_scores)]
    
    print("Average importance score:", np.mean(importance_scores))
    
    return importance_scores


def calculate_importance_scores_for_cluster(cluster_frames, detected_objects_per_frame):
    """Here the importance for each frame based on objects in frame

    Args:
        cluster_dict (_type_): _description_
        detected_objects_per_frame (_type_): _description_

    Returns:
        _type_: _description_
    """
    # Initialize importance scores for the cluster
    importance_scores = []

    # Calculate object-based importance for each frame in the cluster
    for frame_index in cluster_frames:
        # Calculate object importance
        objects = detected_objects_per_frame[frame_index]
        object_importance = len(objects) if objects else 0

        # Add object importance to the list
        importance_scores.append(object_importance)

    # Normalize the importance scores
    max_importance = max(importance_scores, default=1)
    if max_importance == 0:
        max_importance = 1  # To prevent division by zero

    normalized_importance_scores = [score / max_importance for score in importance_scores]

    return normalized_importance_scores


### Map labels with frames

In [161]:
def map_frames_to_labels_with_indices(frames, labels):
    label_frame_dict = {}
    for label, (frame_index, frame) in zip(labels, enumerate(frames)):
        if label not in label_frame_dict:
            label_frame_dict[label] = []
        label_frame_dict[label].append((frame_index, frame))
    return label_frame_dict


# Code to create summary video

#### Variables

In [101]:
annotation_path='datasets/ydata-tvsum50-v1_1/data/ydata-tvsum50-anno.tsv'
info_path='datasets/ydata-tvsum50-v1_1/data/ydata-tvsum50-info.tsv'

In [102]:
video_path='datasets/ydata-tvsum50-v1_1/video/'
summary_video_path='datasets/summary_videos/'

In [103]:
ground_truth_path='datasets/ydata-tvsum50-v1_1/ground_truth/ydata-tvsum50.mat'

##### Get the list of the videos in the folder

In [104]:
video_list = [video for video in os.listdir(video_path) if video.endswith('.mp4')]  # List comprehension

##### Yolo Model

In [105]:
yolo_model = cv2.dnn.readNetFromDarknet('yolo/yolov3.cfg', 'yolo/yolov3.weights') #type:ignore

with open("yolo/coco.names", "r") as f:
    classes = [line.strip() for line in f.readlines()]

In [106]:
frame_rate = 30  # or whatever your frame rate is
frame_duration = 1 / frame_rate  # Duration of each frame in seconds
capacity = 15  # 15 seconds summary

#### Function for videoSummarizion

In [202]:

def videoSumm(annotation_path=None, info_path=None, video_path=None, summary_video_path=None,video_list=None):
    for video in video_list: #type: ignore
        
        # Kmeans clustering
        # frames,labels = MultiModalKMeans(video_path+video, annotation_path, info_path,num_clusters=11)

        """ USE THIS IN CASE OF CREATING VIDEO FOR EACH CLUSTER"""
        cluster_dict = map_frames_to_labels_with_indices(frames, labels)

        scores = []
        for index, cluster in cluster_dict.items():
            print("Cluster:", index)
            # Extract just the frame indices for importance score calculation
            cluster_frame_indices = [frame_index for frame_index, _ in cluster]

            # Calculate importance scores for this cluster
            importance = calculate_importance_scores_for_cluster(cluster_frame_indices, objects) # Adjust your function as necessary

            # Create a list of durations, one for each frame in the cluster
            durations = [frame_duration] * len(importance)

            # Apply the knapsack algorithm
            summary_indices = knapsack_for_video_summary(importance, durations, capacity)

            # Select Representative Frames (extracting frame part from each tuple)
            summary_frames = [cluster[i][1] for i in summary_indices]  # [1] extracts the frame from (frame_index, frame)

            # Create Summary Video
            create_video_from_frames(summary_frames, f"{summary_video_path}{index}-{video}", 30)

            # Extract original indices for evaluation
            original_indices_for_eval = [cluster[i][0] for i in summary_indices]  # [0] extracts the original frame index

            # Evaluate
            ev = Evaluation(ground_truth_path, original_indices_for_eval, video.split('.')[0])
            scores.append((str(index),) + ev)
            print()

        # Print the table of scores
        from tabulate import tabulate
        scores.sort(key=lambda x: x[3], reverse=True)
        print(tabulate(scores, headers=['videoID','Threshold','Precision', 'Recall', 'F1', 'Avg_importance', 'Max_importance', 'Prop_high_importance'], tablefmt='fancy_grid'))
        return
    
        """
        USE THIS IN CASE OF ONE CLUSTER
        importance=calculate_importance_scores(labels,objects) #type: ignore
                
        # Create a list of durations, one for each frame
        durations = [frame_duration] * len(importance)
        
        # Apply the knapsack algorithm
        summary_indices = knapsack_for_video_summary(importance, durations, capacity)

        # Extract summary frames based on 'summary_indices'
        summary_frames = [frames[i] for i in summary_indices] #type: ignore
        
        create_video_from_frames(summary_frames, summary_video_path+video,30) # Create a video from the selected frames
        
        Evaluation(ground_truth_path,summary_indices,video.split('.')[0])
        break
        """

### Function for object detection

### Code for dynamic annotation and video summarization

In [205]:
videoSumm(annotation_path, info_path, video_path, summary_video_path, video_list)

Cluster: 6
threshold 2
precision: 0.43777777777777777
recall: 0.061949685534591195
f1: 0.10853994490358128
avg_importance: 1.9233333333333333
max_importance: 2.4
prop_high_importance: 0.43777777777777777

Cluster: 2
threshold 3
precision: 0.16793893129770993
recall: 0.10476190476190476
f1: 0.12903225806451613
avg_importance: 2.613740458015267
max_importance: 3.0
prop_high_importance: 0.16793893129770993

Cluster: 9
threshold 2
precision: 1.0
recall: 0.03270440251572327
f1: 0.06333739342265529
avg_importance: 2.6024038461538463
max_importance: 2.7
prop_high_importance: 1.0

Cluster: 8
threshold 3
precision: 0.8212290502793296
recall: 0.7
f1: 0.7557840616966581
avg_importance: 3.061312849162012
max_importance: 3.35
prop_high_importance: 0.8212290502793296

Cluster: 10
threshold 2
precision: 0.1927710843373494
recall: 0.005031446540880503
f1: 0.009806926141587496
avg_importance: 1.9367469879518076
max_importance: 2.45
prop_high_importance: 0.1927710843373494

Cluster: 3
No frames to creat

# Results

## After Video call, using gold_standard

Try | Precision | Recall | Weighted F1 | Avg Importance | Max Importance | Proportion High Importance | Clusters |
|---|-----------|--------|-------------|----------------|----------------|----------------------------|----------|
|1  | 1.0       | 0.40379050925925924  | 0.5752859940224|1.86316732353995|3.05| 0.02149767108563239   | 9        |
|2  | 1.0       |0.3392650462962963    |0.50664362104353|1.89402985074626|2.85|0.4597014925373134     |11        |

## Using Only visual_features

### All results was based on frame extraction with frame_rate=25

#### Results using sqrt(N) = 16 Clusters

| Annotator | Precision | Recall | Weighted F1 | Avg Importance | Max Importance | Proportion High Importance |
|-----------|-----------|--------|-------------|----------------|----------------|----------------------------|
| 2         | 0.941     | 0.063  | 0.345       | 3.338          | 5              | 0.131                      |
| 8         | 1.0       | 0.065  | 0.339       | 4.0            | 4              | 1.0                        |
| 12        | 0.766     | 0.052  | 0.337       | 2.111          | 3              | 0.131                      |
| 10        | 1.0       | 0.065  | 0.332       | 3.507          | 4              | 0.507                      |
| 0         | 0.469     | 0.031  | 0.292       | 1.735          | 3              | 0.131                      |
| 1         | 0.473     | 0.03   | 0.274       | 1.394          | 2              | 0.394                      |
| 9         | 0.0       | 0.0    | 0.222       | 1.0            | 1              | 1.0                        |
| 3         | 1.0       | 0.037  | 0.071       | 2.511          | 4              | 0.019                      |
| 4         | 1.0       | 0.037  | 0.071       | 1.0            | 1              | 1.0                        |
| 5         | 1.0       | 0.037  | 0.071       | 2.0            | 2              | 1.0                        |
| 6         | 1.0       | 0.037  | 0.071       | 1.393          | 4              | 0.131                      |
| 7         | 1.0       | 0.037  | 0.071       | 1.0            | 1              | 1.0                        |
| 11        | 1.0       | 0.037  | 0.071       | 3.617          | 5              | 0.131                      |
| 13        | 1.0       | 0.037  | 0.071       | 1.0            | 1              | 1.0                        |
| 14        | 1.0       | 0.037  | 0.071       | 1.0            | 1              | 1.0                        |
| 15        | 1.0       | 0.037  | 0.071       | 4.0            | 4              | 1.0                        |
| 16        | 1.0       | 0.037  | 0.071       | 1.0            | 1              | 1.0                        |
| 17        | 1.0       | 0.037  | 0.071       | 2.0            | 2              | 1.0                        |
| 18        | 1.0       | 0.037  | 0.071       | 1.0            | 1              | 1.0                        |
| 19        | 1.0       | 0.037  | 0.071       | 2.0            | 2              | 1.0                        |


#### Results using 11 Clusters

| Annotator | Precision | Recall | Weighted F1 | Avg Importance | Max Importance | Proportion High Importance |
|-----------|-----------|--------|-------------|----------------|----------------|----------------------------|
| 2         | 0.941     | 0.063  | 0.345       | 3.315          | 5              | 0.117                      |
| 8         | 1.0       | 0.065  | 0.339       | 4.0            | 4              | 1.0                        |
| 12        | 0.766     | 0.052  | 0.337       | 2.099          | 3              | 0.117                      |
| 10        | 1.0       | 0.065  | 0.332       | 3.499          | 4              | 0.499                      |
| 0         | 0.469     | 0.031  | 0.292       | 1.718          | 3              | 0.117                      |
| 1         | 0.473     | 0.03   | 0.274       | 1.396          | 2              | 0.396                      |
| 9         | 0.0       | 0.0    | 0.222       | 1.0            | 1              | 1.0                        |
| 3         | 1.0       | 0.037  | 0.071       | 2.516          | 4              | 0.018                      |
| 4         | 1.0       | 0.037  | 0.071       | 1.0            | 1              | 1.0                        |
| 5         | 1.0       | 0.037  | 0.071       | 2.0            | 2              | 1.0                        |
| 6         | 1.0       | 0.037  | 0.071       | 1.355          | 4              | 0.117                      |
| 7         | 1.0       | 0.037  | 0.071       | 1.0            | 1              | 1.0                        |
| 11        | 1.0       | 0.037  | 0.071       | 3.595          | 5              | 0.117                      |
| 13        | 1.0       | 0.037  | 0.071       | 1.0            | 1              | 1.0                        |
| 14        | 1.0       | 0.037  | 0.071       | 1.0            | 1              | 1.0                        |
| 15        | 1.0       | 0.037  | 0.071       | 4.0            | 4              | 1.0                        |
| 16        | 1.0       | 0.037  | 0.071       | 1.0            | 1              | 1.0                        |
| 17        | 1.0       | 0.037  | 0.071       | 2.0            | 2              | 1.0                        |
| 18        | 1.0       | 0.037  | 0.071       | 1.0            | 1              | 1.0                        |
| 19        | 1.0       | 0.037  | 0.071       | 2.0            | 2              | 1.0                        |


#### Results using 7 Clusters

| Annotator | Precision | Recall | Weighted F1 | Avg Importance | Max Importance | Proportion High Importance |
|-----------|-----------|--------|-------------|----------------|----------------|----------------------------|
| 2         | 0.941     | 0.063  | 0.345       | 3.351          | 5              | 0.134                      |
| 8         | 1.0       | 0.065  | 0.339       | 4.0            | 4              | 1.0                        |
| 12        | 0.766     | 0.052  | 0.337       | 2.115          | 3              | 0.134                      |
| 10        | 1.0       | 0.065  | 0.332       | 3.504          | 4              | 0.504                      |
| 0         | 0.469     | 0.031  | 0.292       | 1.744          | 3              | 0.134                      |
| 1         | 0.473     | 0.03   | 0.274       | 1.386          | 2              | 0.386                      |
| 9         | 0.0       | 0.0    | 0.222       | 1.0            | 1              | 1.0                        |
| 3         | 1.0       | 0.037  | 0.071       | 2.511          | 4              | 0.019                      |
| 4         | 1.0       | 0.037  | 0.071       | 1.0            | 1              | 1.0                        |
| 5         | 1.0       | 0.037  | 0.071       | 2.0            | 2              | 1.0                        |
| 6         | 1.0       | 0.037  | 0.071       | 1.405          | 4              | 0.134                      |
| 7         | 1.0       | 0.037  | 0.071       | 1.0            | 1              | 1.0                        |
| 11        | 1.0       | 0.037  | 0.071       | 3.615          | 5              | 0.134                      |
| 13        | 1.0       | 0.037  | 0.071       | 1.0            | 1              | 1.0                        |
| 14        | 1.0       | 0.037  | 0.071       | 1.0            | 1              | 1.0                        |
| 15        | 1.0       | 0.037  | 0.071       | 4.0            | 4              | 1.0                        |
| 16        | 1.0       | 0.037  | 0.071       | 1.0            | 1              | 1.0                        |
| 17        | 1.0       | 0.037  | 0.071       | 2.0            | 2              | 1.0                        |
| 18        | 1.0       | 0.037  | 0.071       | 1.0            | 1              | 1.0                        |
| 19        | 1.0       | 0.037  | 0.071       | 2.0            | 2              | 1.0                        |

### All results was based on frame extraction with frame_rate=1

#### Using sqrt(N) = 83 clusters
| Annotator | Precision | Recall | Weighted F1 | Avg Importance | Max Importance | Proportion High Importance |
|-----------|-----------|--------|-------------|----------------|----------------|----------------------------|
|     12    |   0.587   |  0.053 |    0.332    |     2.27       |       5        |          0.091             |
|     2     |   0.622   |  0.055 |    0.327    |     1.839      |       4        |          0.109             |
|     8     |   0.434   |  0.038 |    0.294    |     1.801      |       5        |          0.041             |
|     10    |   0.463   |  0.04  |    0.29     |     1.894      |       4        |          0.111             |
|     1     |   0.522   |  0.044 |    0.289    |     1.604      |       5        |          0.021             |
|     0     |   0.246   |  0.022 |    0.273    |     1.378      |       3        |          0.132             |
|     9     |   0.384   |  0.032 |    0.263    |     1.575      |       5        |          0.026             |
|     3     |   1       |  0.049 |    0.094    |     2.279      |       5        |          0.114             |
|     4     |   1       |  0.049 |    0.094    |     2.282      |       5        |          0.114             |
|     5     |   1       |  0.049 |    0.094    |     1.267      |       2        |          0.267             |
|     6     |   1       |  0.049 |    0.094    |     2.525      |       4        |          0.282             |
|     7     |   1       |  0.049 |    0.094    |     2.428      |       5        |          0.111             |
|     11    |   1       |  0.049 |    0.094    |     2.144      |       4        |          0.138             |
|     13    |   1       |  0.049 |    0.094    |     1.358      |       3        |          0.111             |
|     14    |   1       |  0.049 |    0.094    |     1.443      |       4        |          0.091             |
|     15    |   1       |  0.049 |    0.094    |     1.56       |       5        |          0.021             |
|     16    |   1       |  0.049 |    0.094    |     1.117      |       2        |          0.117             |
|     17    |   1       |  0.049 |    0.094    |     1.299      |       4        |          0.091             |
|     18    |   1       |  0.049 |    0.094    |     2.563      |       5        |          0.111             |
|     19    |   1       |  0.049 |    0.094    |     1.472      |       4        |          0.111             |



#### Using 10 clusters
| Annotator | Precision | Recall | Weighted F1 | Avg Importance | Max Importance | Proportion High Importance |
|-----------|-----------|--------|-------------|----------------|----------------|----------------------------|
|     8     |   0.856   | 0.583  |    0.706    |     2.486      |       5        |          0.009             |
|     3     |   1       | 0.385  |    0.556    |     2.039      |       5        |          0.017             |
|     4     |   1       | 0.385  |    0.556    |     2.124      |       5        |          0.017             |
|     5     |   1       | 0.385  |    0.556    |     1.505      |       3        |          0.003             |
|     6     |   1       | 0.385  |    0.556    |     2.135      |       5        |          0.027             |
|     7     |   1       | 0.385  |    0.556    |     1.85       |       5        |          0.014             |
|     11    |   1       | 0.385  |    0.556    |     2.12       |       5        |          0.023             |
|     13    |   1       | 0.385  |    0.556    |     1.337      |       4        |          0.008             |
|     14    |   1       | 0.385  |    0.556    |     1.738      |       5        |          0.084             |
|     15    |   1       | 0.385  |    0.556    |     1.756      |       5        |          0.003             |
|     16    |   1       | 0.385  |    0.556    |     1.363      |       4        |          0.007             |
|     17    |   1       | 0.385  |    0.556    |     1.571      |       4        |          0.03              |
|     18    |   1       | 0.385  |    0.556    |     2.139      |       5        |          0.014             |
|     19    |   1       | 0.385  |    0.556    |     1.479      |       4        |          0.037             |
|     12    |   0.624   | 0.439  |    0.543    |     2.105      |       5        |          0.039             |
|     2     |   0.538   | 0.373  |    0.468    |     1.975      |       5        |          0.045             |
|     10    |   0.389   | 0.262  |    0.337    |     1.718      |       5        |          0.008             |
|     1     |   0.391   | 0.259  |    0.33     |     1.784      |       5        |          0.051             |
|     0     |   0.331   | 0.226  |    0.299    |     1.646      |       5        |          0.084             |
|     9     |   0.302   | 0.197  |    0.253    |     1.72       |       5        |          0.132             |



#### Using 9 clusters 1.1m

| Annotator | Precision | Recall | Weighted F1 | Avg Importance | Max Importance | Proportion High Importance |
|-----------|-----------|--------|-------------|----------------|----------------|----------------------------|
|     8     |   0.837   |  0.487 |    0.648    |     2.706      |       5        |          0.158             |
|     3     |     1     |  0.33  |    0.496    |     1.934      |       5        |          0.024             |
|     4     |     1     |  0.33  |    0.496    |     1.922      |       5        |          0.024             |
|     5     |     1     |  0.33  |    0.496    |     1.313      |       2        |          0.313             |
|     6     |     1     |  0.33  |    0.496    |     2.235      |       5        |          0.057             |
|     7     |     1     |  0.33  |    0.496    |     1.859      |       5        |          0.017             |
|     11    |     1     |  0.33  |    0.496    |     1.998      |       4        |          0.113             |
|     13    |     1     |  0.33  |    0.496    |     1.187      |       3        |          0.025             |
|     14    |     1     |  0.33  |    0.496    |     1.324      |       4        |          0.026             |
|     15    |     1     |  0.33  |    0.496    |     1.602      |       5        |          0.026             |
|     16    |     1     |  0.33  |    0.496    |     1.106      |       4        |          0.008             |
|     17    |     1     |  0.33  |    0.496    |     1.293      |       4        |          0.026             |
|     18    |     1     |  0.33  |    0.496    |     2.089      |       5        |          0.017             |
|     19    |     1     |  0.33  |    0.496    |     1.64       |       5        |          0.026             |
|     12    |   0.536   |  0.323 |    0.462    |     1.965      |       5        |          0.026             |
|     2     |   0.416   |  0.247 |    0.373    |     1.655      |       5        |          0.026             |
|     1     |   0.388   |  0.22  |    0.33     |     1.812      |       5        |          0.053             |
|     10    |   0.344   |  0.198 |    0.308    |     1.519      |       4        |          0.017             |
|     0     |   0.249   |  0.145 |    0.252    |     1.433      |       3        |          0.184             |
|     9     |   0.283   |  0.158 |    0.252    |     1.442      |       5        |          0.045             |


#### Using 7 clusters
| Annotator | Precision | Recall | Weighted F1 | Avg Importance | Max Importance | Proportion High Importance |
|-----------|-----------|--------|-------------|----------------|----------------|----------------------------|
|     8     |   0.77    |  0.68  |    0.706    |     2.657      |       5        |          0.104             |
|     3     |   1       |  0.5   |    0.666    |     2.002      |       5        |          0.017             |
|     4     |   1       |  0.5   |    0.666    |     1.876      |       5        |          0.017             |
|     5     |   1       |  0.5   |    0.666    |     1.444      |       3        |          0.002             |
|     6     |   1       |  0.5   |    0.666    |     1.954      |       5        |          0.028             |
|     7     |   1       |  0.5   |    0.666    |     1.887      |       5        |          0.017             |
|     11    |   1       |  0.5   |    0.666    |     1.991      |       5        |          0.017             |
|     13    |   1       |  0.5   |    0.666    |     1.518      |       3        |          0.155             |
|     14    |   1       |  0.5   |    0.666    |     1.767      |       5        |          0.065             |
|     15    |   1       |  0.5   |    0.666    |     2.054      |       5        |          0.033             |
|     16    |   1       |  0.5   |    0.666    |     1.605      |       4        |          0.12              |
|     17    |   1       |  0.5   |    0.666    |     1.603      |       4        |          0.017             |
|     18    |   1       |  0.5   |    0.666    |     1.88       |       5        |          0.017             |
|     19    |   1       |  0.5   |    0.666    |     1.865      |       5        |          0.017             |
|     12    |   0.623   |  0.569 |    0.577    |     2.119      |       5        |          0.03              |
|     2     |   0.468   |  0.421 |    0.415    |     1.871      |       5        |          0.039             |
|     10    |   0.436   |  0.38  |    0.367    |     1.82       |       4        |          0.07              |
|     1     |   0.432   |  0.371 |    0.355    |     1.893      |       5        |          0.069             |
|     9     |   0.435   |  0.369 |    0.351    |     1.839      |       5        |          0.089             |
|     0     |   0.333   |  0.295 |    0.272    |     1.666      |       5        |          0.065             |


## Combined Features

### All results was based on frame extraction with frame_rate=25

#### Using 16 clusters


#### Using 10 clusters


#### Using             7 clusters



### All results was based on frame extraction with frame_rate=1

#### Using sqrt(N) = 83 clusters

| Annotator | Precision | Recall | Weighted F1 | Avg Importance | Max Importance | Proportion High Importance |
|-----------|-----------|--------|-------------|----------------|----------------|----------------------------|
|     10    |     1     |  0.107 |    0.378    |     2.123      |       3        |          0.123             |
|     9     |     1     |  0.104 |    0.36     |     2          |       2        |          1                 |
|     8     |   0.825   |  0.089 |    0.359    |     2.65       |       3        |          0.825             |
|     0     |   0.683   |  0.074 |    0.339    |     1.683      |       2        |          0.683             |
|     1     |   0.757   |  0.08  |    0.332    |     2.506      |       4        |          0.135             |
|     12    |   0.123   |  0.014 |    0.273    |     1.123      |       2        |          0.123             |
|     2     |   0.123   |  0.014 |    0.265    |     1.123      |       2        |          0.123             |
|     3     |     1     |  0.061 |    0.115    |     1.752      |       3        |          0.243             |
|     4     |     1     |  0.061 |    0.115    |     1          |       1        |          1                 |
|     5     |     1     |  0.061 |    0.115    |     1          |       1        |          1                 |
|     6     |     1     |  0.061 |    0.115    |     1.369      |       4        |          0.123             |
|     7     |     1     |  0.061 |    0.115    |     3          |       3        |          1                 |
|     11    |     1     |  0.061 |    0.115    |     1.733      |       3        |          0.366             |
|     13    |     1     |  0.061 |    0.115    |     1          |       1        |          1                 |
|     14    |     1     |  0.061 |    0.115    |     1          |       1        |          1                 |
|     15    |     1     |  0.061 |    0.115    |     1          |       1        |          1                 |
|     16    |     1     |  0.061 |    0.115    |     3          |       3        |          1                 |
|     17    |     1     |  0.061 |    0.115    |     2          |       2        |          1                 |
|     18    |     1     |  0.061 |    0.115    |     1          |       1        |          1                 |
|     19    |     1     |  0.061 |    0.115    |     1          |       1        |          1                 |


#### Using 10 clusters
| Annotator | Precision | Recall | Weighted F1 | Avg Importance | Max Importance | Proportion High Importance |
|-----------|-----------|--------|-------------|----------------|----------------|----------------------------|
|     3     |     1     |  0.528 |    0.691    |     1.994      |       5        |          0.015             |
|     4     |     1     |  0.528 |    0.691    |     1.925      |       5        |          0.015             |
|     5     |     1     |  0.528 |    0.691    |     1.844      |       4        |          0.115             |
|     6     |     1     |  0.528 |    0.691    |     2.298      |       5        |          0.069             |
|     7     |     1     |  0.528 |    0.691    |     1.762      |       5        |          0.01              |
|     11    |     1     |  0.528 |    0.691    |     1.872      |       5        |          0.016             |
|     13    |     1     |  0.528 |    0.691    |     1.839      |       5        |          0.033             |
|     14    |     1     |  0.528 |    0.691    |     2.005      |       5        |          0.066             |
|     15    |     1     |  0.528 |    0.691    |     2.012      |       5        |          0.016             |
|     16    |     1     |  0.528 |    0.691    |     1.626      |       4        |          0.005             |
|     17    |     1     |  0.528 |    0.691    |     1.941      |       4        |          0.088             |
|     18    |     1     |  0.528 |    0.691    |     2.012      |       5        |          0.01              |
|     19    |     1     |  0.528 |    0.691    |     2.122      |       5        |          0.049             |
|     12    |   0.668   |  0.645 |    0.631    |     2.282      |       5        |          0.066             |
|     8     |   0.683   |  0.638 |    0.629    |     2.319      |       5        |          0.099             |
|     2     |   0.662   |  0.629 |    0.616    |     2.35       |       5        |          0.062             |
|     0     |   0.569   |  0.533 |    0.51     |     2.046      |       5        |          0.066             |
|     9     |   0.558   |  0.499 |    0.475    |     1.978      |       5        |          0.048             |
|     1     |   0.509   |  0.462 |    0.432    |     2.093      |       5        |          0.066             |
|     10    |   0.454   |  0.418 |    0.381    |     2.05       |       5        |          0.088             |


#### Using 9 clusters 2m video

| Annotator | Precision | Recall | Weighted F1 | Avg Importance | Max Importance | Proportion High Importance |
|-----------|-----------|--------|-------------|----------------|----------------|----------------------------|
|     3     |     1     |  0.543 |    0.704    |     2.018      |       5        |          0.032             |
|     4     |     1     |  0.543 |    0.704    |     1.799      |       5        |          0.016             |
|     5     |     1     |  0.543 |    0.704    |     1.828      |       4        |          0.112             |
|     6     |     1     |  0.543 |    0.704    |     2.05       |       5        |          0.049             |
|     7     |     1     |  0.543 |    0.704    |     1.755      |       5        |          0.016             |
|     11    |     1     |  0.543 |    0.704    |     1.982      |       5        |          0.016             |
|     13    |     1     |  0.543 |    0.704    |     1.738      |       4        |          0.048             |
|     14    |     1     |  0.543 |    0.704    |     1.829      |       5        |          0.06              |
|     15    |     1     |  0.543 |    0.704    |     1.899      |       5        |          0.027             |
|     16    |     1     |  0.543 |    0.704    |     1.797      |       4        |          0.111             |
|     17    |     1     |  0.543 |    0.704    |     1.762      |       4        |          0.016             |
|     18    |     1     |  0.543 |    0.704    |     1.954      |       5        |          0.016             |
|     19    |     1     |  0.543 |    0.704    |     1.865      |       5        |          0.014             |
|     12    |   0.619   |  0.615 |    0.582    |     2.158      |       5        |          0.044             |
|     8     |   0.612   |  0.588 |    0.557    |     2.189      |       5        |          0.032             |
|     2     |   0.547   |  0.534 |    0.496    |     1.914      |       5        |          0.03              |
|     10    |   0.554   |  0.525 |    0.487    |     2.226      |       5        |          0.064             |
|     9     |   0.466   |  0.429 |    0.377    |     1.986      |       5        |          0.035             |
|     1     |   0.447   |  0.417 |    0.363    |     1.815      |       5        |          0.051             |
|     0     |   0.372   |  0.358 |    0.298    |     1.76       |       5        |          0.06              |


#### Using 7 clusters 2.4m video

| Annotator | Precision | Recall | Weighted F1 | Avg Importance | Max Importance | Proportion High Importance |
|-----------|-----------|--------|-------------|----------------|----------------|----------------------------|
|     3     |     1     |  0.7   |    0.823    |     1.966      |       5        |          0.025             |
|     4     |     1     |  0.7   |    0.823    |     1.782      |       5        |          0.012             |
|     5     |     1     |  0.7   |    0.823    |     1.692      |       4        |          0.087             |
|     6     |     1     |  0.7   |    0.823    |     2.047      |       5        |          0.052             |
|     7     |     1     |  0.7   |    0.823    |     2.026      |       5        |          0.012             |
|     11    |     1     |  0.7   |    0.823    |     1.771      |       4        |          0.087             |
|     13    |     1     |  0.7   |    0.823    |     1.833      |       5        |          0.025             |
|     14    |     1     |  0.7   |    0.823    |     1.932      |       5        |          0.05              |
|     15    |     1     |  0.7   |    0.823    |     1.78       |       5        |          0.024             |
|     16    |     1     |  0.7   |    0.823    |     1.981      |       4        |          0.086             |
|     17    |     1     |  0.7   |    0.823    |     1.925      |       4        |          0.062             |
|     18    |     1     |  0.7   |    0.823    |     1.852      |       5        |          0.012             |
|     19    |     1     |  0.7   |    0.823    |     1.977      |       5        |          0.037             |
|     9     |   0.648   |  0.767 |    0.603    |     2.051      |       5        |          0.027             |
|     8     |   0.612   |  0.757 |    0.575    |     2.153      |       5        |          0.074             |
|     2     |   0.528   |  0.665 |    0.463    |     1.977      |       5        |          0.025             |
|     10    |   0.521   |  0.636 |    0.437    |     1.993      |       5        |          0.062             |
|     0     |   0.5     |  0.619 |    0.414    |     1.886      |       5        |          0.05              |
|     1     |   0.497   |  0.598 |    0.393    |     2.028      |       5        |          0.062             |
